Kernel/FileSystem/TmpFS.cpp

```diff
@@ -0,0 +1,295 @@
+
+    memcpy(m_content.offset_pointer(offset), buffer, size);
+
+    return size;
+}
+
+InodeIdentifier TmpFSInode::lookup(StringView name)
+{
+    LOCKER(m_lock);
+    ASSERT(is_directory());
+
+    if (name == ".")
+        return identifier();
+    if (name == "..")
+        return m_parent;
+
+    auto it = m_children.find(name);
+    if (it == m_children.end())
+        return {};
+    return it->value.entry.inode;
+}
+
+size_t TmpFSInode::directory_entry_count() const
+{
+    LOCKER(m_lock);
+    ASSERT(is_directory());
+
+    return 2 + m_children.size();
+}
+
+void TmpFSInode::flush_metadata()
+{
+    // We don't really have any metadata that could become dirty.
+    // The only reason we even call set_metadata_dirty() is
+    // to let the watchers know we have updates. Once that is
+    // switched to a different mechanism, we can stop ever marking
+    // our metadata as dirty at all.
+    set_metadata_dirty(false);
+}
+
+KResult TmpFSInode::chmod(mode_t mode)
+{
+    LOCKER(m_lock);
+
+    m_metadata.mode = mode;
+    set_metadata_dirty(true);
+    set_metadata_dirty(false);
+    return KSuccess;
+}
+
+KResult TmpFSInode::chown(uid_t uid, gid_t gid)
+{
+    LOCKER(m_lock);
+
+    m_metadata.uid = uid;
+    m_metadata.gid = gid;
+    set_metadata_dirty(true);
+    set_metadata_dirty(false);
+    return KSuccess;
+}
+
+KResult TmpFSInode::add_child(InodeIdentifier child_id, const StringView& name, mode_t)
+{
+    LOCKER(m_lock);
+    ASSERT(is_directory());
+    ASSERT(child_id.fsid() == fsid());
+
+    String owned_name = name;
+    FS::DirectoryEntry entry = { owned_name.characters(), owned_name.length(), child_id, 0 };
+    RefPtr<Inode> child_tmp = fs().get_inode(child_id);
+    NonnullRefPtr<TmpFSInode> child = static_cast<NonnullRefPtr<TmpFSInode>>(child_tmp.release_nonnull());
+
+    m_children.set(owned_name, { entry, move(child) });
+    set_metadata_dirty(true);
+    set_metadata_dirty(false);
+    return KSuccess;
+}
+
+KResult TmpFSInode::remove_child(const StringView& name)
+{
+    LOCKER(m_lock);
+    ASSERT(is_directory());
+
+    auto it = m_children.find(name);
+    if (it == m_children.end())
+        return KResult(-ENOENT);
+    m_children.remove(it);
+    set_metadata_dirty(true);
+    set_metadata_dirty(false);
+    return KSuccess;
+}
+
+int TmpFSInode::set_atime(time_t time)
+{
+    LOCKER(m_lock);
+
+    m_metadata.atime = time;
+    set_metadata_dirty(true);
+    set_metadata_dirty(false);
+    return KSuccess;
+}
+
+int TmpFSInode::set_ctime(time_t time)
+{
+    LOCKER(m_lock);
+
+    m_metadata.ctime = time;
+    set_metadata_dirty(true);
+    set_metadata_dirty(false);
+    return KSuccess;
+}
+
+int TmpFSInode::set_mtime(time_t time)
+{
+    LOCKER(m_lock);
+
+    m_metadata.mtime = time;
+    set_metadata_dirty(true);
+    set_metadata_dirty(false);
+    return KSuccess;
+}
+
+void TmpFSInode::one_ref_left()
+{
+    // Destroy ourselves.
+    static_cast<TmpFS&>(fs()).unregister_inode(identifier());
```

Comment:
Let's add an override for `fs()` in `TmpFSInode` that does the casting for us:
```cpp
TmpFS& fs() { return static_cast<TmpFS&>(Inode::fs()); }
const TmpFS& fs() const { return static_cast<const TmpFS&>(Inode::fs()); }
```

That way we'll always get the right type when we call `fs()` inside `TmpFSInode`